docs(fx112-css-overflow): Add note for 'overlay' and update other descriptions (#25749)
mdn/content: files/en-us/web/css/overflow/index.md

- `scroll`
  - : Overflow content is clipped at the content's padding box, as with `hidden`. Unlike `hidden`, overflow content can be scrolled into view in the {{glossary("Scrollport", "scrollport")}} (the element box is a {{glossary("scroll container")}}). User agents always display scroll bars in both horizontal and vertical directions, whether or not any content is overflowing or clipped. The use of this keyword, therefore, can prevent scroll bars from appearing and disappearing as content changes. Printers may still print overflowing content.
- `auto`
  - : Overflow content is clipped at the content's padding box, as with `hidden`, and overflow content can be scrolled into view, as with `scroll`. User agents display scroll bars _only if_ the content is overflowing and hide scroll bars by default. If content fits inside the padding box, it looks the same as `visible` but still establishes a new [block formatting context](/en-US/docs/Web/Guide/CSS/Block_formatting_context).

> **Note:** The keyword `overlay` is a legacy value alias for `auto` and behaves the same as `auto`.

> #### Mozilla extensions

Reviewer: Drive by comments: This page is supposed to be limited to the `overflow` property. So, yes, mozilla extensions should be removed (and quite possibly shouldn't have been here). 

We're getting into the weeds of formatting context, clip margin, scroll container, etc. Values are generally very brief with what the value does in terms of the property only, not how it interacts with other properties and concepts. This content is better suited in examples, # #see also, ## performance considerations, and similar.

overflow: unset;
```

The `overflow` property is specified as one or two keywords chosen from the list of options below. If only one keyword is specified, both `overflow-x` and `overflow-y` are set to the same value. If two keywords are specified, the first value applies to `overflow-x` in the horizontal direction and the second one applies to `overflow-y` in the vertical direction.

### Values

- `visible`
  - : Overflow content is not clipped and may be visible outside the element's padding box. The element box is not a {{glossary("scroll container")}}. This is the default value of the `overflow` property.
- `hidden`

>   - : Overflow content is clipped at the element's padding box. There are no scroll bars, the clipped content is not visible (i.e., clipped content is hidden), but the content still exists. User agents do not add scroll bars and also do not allow users to view the content outside the clipped region by actions such as dragging on a touch screen or using the scroll wheel on a mouse. The content _can_ be scrolled programmatically (for example, by setting the value of the {{domxref("Element.scrollLeft", "scrollLeft")}} property or the {{domxref("Element.scrollTo", "scrollTo()")}} method), in which case. The element box is a scroll container.

Reviewer: ```suggestion
  - : Overflow content is clipped at the element's padding box. There are no scroll bars and the clipped content is not visible (i.e., clipped content is hidden), but the content still exists. User agents do not add scroll bars and also do not allow users to view the content outside the clipped region by actions such as dragging on a touch screen or using the scroll wheel on a mouse. The content _can_ be scrolled programmatically (for example, by setting the value of the {{domxref("Element.scrollLeft", "scrollLeft")}} property or the {{domxref("Element.scrollTo", "scrollTo()")}} method). The element box is a scroll container.
```

- `clip`

>   - : Overflow content is clipped at the element's _overflow clip edge_ that is defined using the [`overflow-clip-margin`](/en-US/docs/Web/CSS/overflow-clip-margin) property. As a result, content overflows the element's padding box by the {{cssxref("&lt;length&gt;")}} value of `overflow-clip-margin` or by `0px` if not set. Overflow content outside the clipped region is not visible, user agents do not add a scroll bar, and programmatic scrolling is also not supported. No new [formatting context](/en-US/docs/Web/Guide/CSS/Block_formatting_context) can be created. To establish a formatting context, use `overflow: clip` along with {{cssxref("display", "display: flow-root", "#flow-root")}}. The element box is not a scroll container.

Reviewer: ```suggestion
  - : Overflow content is clipped at the element's _overflow clip edge_ that is defined using the [`overflow-clip-margin`](/en-US/docs/Web/CSS/overflow-clip-margin) property. As a result, content overflows the element's padding box by the {{cssxref("&lt;length&gt;")}} value of `overflow-clip-margin` or by `0px` if not set. Overflow content outside the clipped region is not visible, user agents do not add a scroll bar, and programmatic scrolling is also not supported. No new [formatting context](/en-US/docs/Web/Guide/CSS/Block_formatting_context) is created. To establish a formatting context, use `overflow: clip` along with {{cssxref("display", "display: flow-root", "#flow-root")}}. The element box is not a scroll container.
```

- `scroll`

>   - : Overflow content is clipped at the element's padding box, and overflow content can be scrolled into view using scroll bars. User agents always display scroll bars in both horizontal and vertical directions, whether or not any content is overflowing or clipped. The use of this keyword, therefore, can prevent scroll bars from appearing and disappearing as content changes. Printers may still print overflowing content. The element box is a scroll container.

Reviewer: ```suggestion
  - : Overflow content is clipped at the element's padding box, and overflow content can be scrolled into view using scroll bars. User agents display scroll bars in the direction on which this value is set, in both horizontal and vertical directions if only one value is set, whether or not any content is overflowing or clipped. The use of this keyword, therefore, can prevent scroll bars from appearing and disappearing as content changes. Printers may still print overflowing content. The element box is a scroll container.
```

> **Note:** The keyword value `overlay` is a legacy value alias for `auto`. With `overlay`, the scroll bars are drawn on top of the content instead of taking up space.

## Description

Overflow options include hiding overflowing content, enabling scroll bars to view overflow content or displaying the content flowing out of an element box into the surrounding area, and combinations there of.

The following nuances should be kept in mind while using the various keywords for `overflow`:

- Specifying a value other than `visible` (the default) or `clip` for `overflow` creates a new [block formatting context](/en-US/docs/Web/Guide/CSS/Block_formatting_context). This is necessary for technical reasons; if a float intersects with a scrolling element, it would forcibly rewrap the content after each scroll step, leading to a slow scrolling experience.
- For an `overflow` setting to create the desired effect, the block-level element must have either a set height (`height` or `max-height`) or `white-space` set to `nowrap`.
- Setting one axis to `visible` (the default) while setting the other to a _different_ value results in `visible` behaving as `auto`.

> - The JavaScript {{domxref("Element.scrollTop")}} property may be used to scroll an HTML element even when `overflow` is set to `hidden`.

Reviewer: ```suggestion
- The JavaScript {{domxref("Element.scrollTop")}} property may be used to scroll scroll containers, including when `overflow` is set to `hidden`.
```

Reviewer: does not work on clip and visible

overflow: unset;
```

> The `overflow` property is specified as one or two keywords chosen from the list of options below. If only one keyword is specified, both `overflow-x` and `overflow-y` are set to the same value. If two keywords are specified, the first value applies to `overflow-x` and the second one applies to `overflow-y`, always in that order.

Reviewer 2: Redundant.

```suggestion
The `overflow` property is specified as one or two keywords chosen from the list of options below. If only one keyword is specified, both `overflow-x` and `overflow-y` are set to the same value. If two keywords are specified, the first value applies to `overflow-x` and the second one applies to `overflow-y`.
```

### Values

- `visible`
  - : Overflow content is not clipped and may be visible outside content's padding box. This is the default value. In this case, the element is not a {{glossary("scroll container")}}.
- `hidden`

>   - : Overflow content is clipped at the content's padding box. Overflow content outside the clipped region is not visible. User agents do not add a scroll bar and also do not allow users to view the content outside the clipped region by actions such as dragging on a touch screen or using the scroll wheel on a mouse. The content _can_ still be scrolled programmatically (for example, by setting the value of the {{domxref("Element.scrollLeft", "scrollLeft")}} property or the {{domxref("Element.scrollTo", "scrollTo()")}} method), in which case, the element box is a {{glossary("scroll container")}}.

Author: > Test page: https://codepen.io/estelle/pen/mdzeWqN
> You can tab thru the links and they get brought into view.

When I tab through, the element boxes with `overflow` set to `hidden`, `clip`, and `visible` get skipped. Only `scroll`, `auto`, and `overlay` get focus. And so I am not able to validate this additional text:

> "tabbing to hidden elements in the page's tabbing sequence does scroll the focused content into view."

Is this correct - `hidden` content cannot be tabbed into view, but we're still saying it is a scroll container because there is a possibility to add scroll bars (but you're not able to add scroll bars using JS)?

> do we want to link every occurence of scroll container?

Nope, too many of them in a short space. I've removed the extra ones. Thanks for pointing out.

>   - : Overflow content is clipped at the content's padding box. Overflow content outside the clipped region is not visible. User agents do not add a scroll bar and also do not allow users to view the content outside the clipped region by actions such as dragging on a touch screen or using the scroll wheel on a mouse. The content _can_ still be scrolled programmatically though (for example, by setting the value of the {{domxref("Element.scrollLeft", "scrollLeft")}} property or the {{domxref("Element.scrollTo", "scrollTo()")}} method), in which case, the element box is a {{glossary("scroll container")}}.

Reviewer 2: YOu don't need "though" here, though does no particular harm.
```suggestion
  - : Overflow content is clipped at the content's padding box. Overflow content outside the clipped region is not visible. User agents do not add a scroll bar and also do not allow users to view the content outside the clipped region by actions such as dragging on a touch screen or using the scroll wheel on a mouse. The content _can_ still be scrolled programmatically (for example, by setting the value of the {{domxref("Element.scrollLeft", "scrollLeft")}} property or the {{domxref("Element.scrollTo", "scrollTo()")}} method), in which case, the element box is a {{glossary("scroll container")}}.
```

- `clip`
  - : Overflow content is clipped, as with `hidden`, but at the element's _overflow clip edge_ (defined using the [`overflow-clip-margin`](/en-US/docs/Web/CSS/overflow-clip-margin) property). Similar to `hidden`, overflow content outside the clipped region is not visible with the `clip` keyword, and user agents do not add a scroll bar. However, unlike `hidden`, the `clip` keyword does not allow adding support for scrolling programmatically. Therefore, with `overflow: clip`, an element box is never a {{glossary("scroll container")}}.

> **Note:** `overflow: clip` prevents a new [formatting context](/en-US/docs/Web/Guide/CSS/Block_formatting_context). If you want the element box to also establish a formatting context, use `overflow: clip` along with {{cssxref("display", "display: flow-root", "#flow-root")}}.

- `scroll`

>   - : Overflow content is clipped at the content's padding box, as with `hidden`. Unlike `hidden`, overflow content can be scrolled into view in the {{glossary("Scrollport", "scrollport")}} (the element box is a {{glossary("scroll container")}}). User agents always display scroll bars in both horizontal and vertical directions, whether or not any content is overflowing or clipped. The use of this keyword, therefore, can prevent scroll bars from appearing and disappearing as content changes. Printers may still print overflowing content.

Author: Not sure about the validity of this part - "though they may be hidden by default, becoming visible only during user interaction". I thought scroll bars are always visible in the case of `scroll`.

I am inclined to keep this bit because it conveys when somebody might want to use this setting: "The use of this keyword, therefore, can prevent scroll bars from appearing and disappearing as content changes. "

- `auto`
  - : Overflow content is clipped at the content's padding box, as with `hidden`, and overflow content can be scrolled into view, as with `scroll`. User agents display scroll bars _only if_ the content is overflowing and hide scroll bars by default. If content fits inside the padding box, it looks the same as `visible` but still establishes a new [block formatting context](/en-US/docs/Web/Guide/CSS/Block_formatting_context).

> > **Note:** The keyword `overlay` is a legacy value alias for `auto` and behaves the same as `auto`.

Author: The page originally did have an inline deprecated sign next to `overlay` (because it was not supported in Firefox) and it [was listed as a separate value](https://developer.mozilla.org/en-US/docs/Web/CSS/overflow#values).

It is not deprecated now. I don't want to list it as a separate value because:
- The syntax provides five options and I want the list under the "Values" section to match that.
- In Chrome DevTools, `overflow` options do list `overlay`. In Firefox DevTools, `overlay` is not listed as an option for `overflow`, but it is now supported in Firefox (via the [bug fix](https://bugzilla.mozilla.org/show_bug.cgi?id=1817189)).

> Also, this value is NOT the same as auto.

Removed "and behaves the same as auto." and brought back the earlier "but with the scrollbars drawn on top of content instead of taking up space."

## Description

Overflow options include hiding overflowing content, showing scroll bars to view overflow content, or displaying the content flowing out of an element box into the surrounding area.

The following nuances should be kept in mind while using the various keywords for `overflow`:

- Specifying a value other than `visible` (the default) or `clip` for `overflow` creates a new [block formatting context](/en-US/docs/Web/Guide/CSS/Block_formatting_context). This is necessary for technical reasons; if a float intersects with a scrolling element, it would forcibly rewrap the content after each scroll step, leading to a slow scrolling experience.
- For an `overflow` setting to create the desired effect, the block-level element must have either a set height (`height` or `max-height`) or `white-space` set to `nowrap`.
- Setting one axis to `visible` (the default) while setting the other to a _different_ value results in `visible` behaving as `auto`.

> - The JavaScript {{domxref("Element.scrollTop")}} property may be used to scroll an HTML element even when `overflow` is set to `hidden`.

Reviewer: https://codepen.io/estelle/live/mdzeWqN (has focusable content)
https://codepen.io/estelle/live/rNqOYNN (doesn't)

  </p>
</div>

<div>
  <code>clip</code>

>   <p class="clip">

Reviewer: Did you want to include an inspirational quote, such as "I've learned that people will forget what you said, people will forget what you did, but people will never forget how you made them feel. -- Maya Angelou"

Author: Done. Love the quote. Put her name in the beginning so that it is always visible even if the content is hidden ;)

#### Result

{{EmbedLiveSample("Demonstrating results of various overflow keywords", "600", "400")}}

## Accessibility concerns

A scrolling content area cannot be scrolled by a keyboard-only user, with the exception of users on Firefox (which makes the container keyboard focusable by default).

> As a developer, to allow non-Firefox keyboard-only users to scroll the container, you will need to give it a [`tabindex`](/en-US/docs/Web/HTML/Global_attributes/tabindex) using `tabindex="0"`. Unfortunately, when a screen reader encounters this tab-stop, they will have no context for what it is and their screen reader will likely announce the entirety of its contents. Giving it an appropriate [WAI-ARIA role](/en-US/docs/Web/Accessibility/ARIA/Roles) (`role="region"`, for example) and an accessible name (via [`aria-label`](/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-label) or [`aria-labelledby`](/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-labelledby)) can mitigate this.

Reviewer: Chrome supports this too. I have no clue what Safari is doing though. Really odd.
